barretenberg/cpp/src/barretenberg/vm/avm/trace/gadgets/poseidon2.cpp

```diff
@@ -61,10 +60,106 @@ std::array<FF, 4> AvmPoseidon2TraceBuilder::poseidon2_permutation(std::array<FF,
     }
 
     // Current state is the output
-    poseidon2_trace.push_back(
-        Poseidon2TraceEntry{ clk, input, current_state, first_ext, interm_round_vals, input_addr, output_addr });
+    return AvmPoseidon2TraceBuilder::Poseidon2TraceEntry{
+        clk, input, current_state, first_ext, interm_round_vals,
+    };
+}
 
-    return current_state;
+std::array<FF, 4> AvmPoseidon2TraceBuilder::poseidon2_permutation(std::array<FF, 4> const& input,
+                                                                  uint32_t clk,
+                                                                  uint32_t input_addr,
+                                                                  uint32_t output_addr)
+{
+    auto entry = gen_poseidon_perm_entry(input, clk);
+    entry.input_addr = input_addr;
+    entry.output_addr = output_addr;
+    entry.is_mem_op = true;
+    poseidon2_trace.push_back(entry);
+
+    return entry.output;
 }
 
+FF AvmPoseidon2TraceBuilder::poseidon2_hash(std::vector<FF> input, uint32_t clk)
+{
+    using Poseidon2 = crypto::Poseidon2<crypto::Poseidon2Bn254ScalarFieldParams>;
+    FF output = Poseidon2::hash(input);
+    // 64 rounds of hashing should be enough (1 << 6 == 64) per full hash
```

Comment:
Maybe let us introduce a constant for this 64 and some explanations (related to sub clock, etc...)